lib/Transforms/GenStateStruct.cpp

```diff
@@ -0,0 +1,1021 @@
+  /// infomation of current module. The struct type is packed defaultly.
+  /// If the module is an extern module which only have a declaration, generate
+  /// opaque struct type for it.
+  LLVM::LLVMStructType createStructType(MLIRContext *ctx, HWModuleOp op,
+                                        InstanceGraph &instGraph,
+                                        SmallVector<PortInfoCollect> &pInfo,
+                                        SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function to create the global struct that is used to store all state
+  /// values and ports infomation.
+  LLVM::GlobalOp createGlobalStruct(Location loc, OpBuilder &builder,
+                                    ModuleOp module, StringRef name,
+                                    LLVM::LLVMStructType type);
+
+  /// The function to initialize the global struct with default value 0, and
+  /// return initialized struct object.
+  void initializeGlobalStruct(OpBuilder &builder, HWModuleOp op,
+                              LLVM::GlobalOp globalOp,
+                              LLVM::LLVMStructType type,
+                              InstanceGraph &instGraph,
+                              SmallVector<PortInfoCollect> &pInfo,
+                              SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function is used to collect all ports info above the given module.
+  /// The collected info: port name, port type, port direction and whether it is
+  /// trigger signal.
+  void collectPortInfo(HWModuleOp op, SmallVector<PortInfoCollect> &pInfo);
+
+  /// The function is used to collect all regs info in the given module.
+  /// The collected info: reg name, reg type and whether it is output reg.
+  void collectRegInfo(HWModuleOp op, SmallVector<RegInfoCollect> &rInfo);
+
+  /// The function to override all modules in InstanceGraph based on a
+  /// depth-first algorithm.
+  void dfsOverrideModules(
+      MLIRContext *ctx,
+      SmallVector<std::pair<mlir::Operation *, LLVM::LLVMStructType>>
+          &typeCollect,
+      SmallVector<std::pair<mlir::Operation *, LLVM::LLVMPointerType>>
+          &externTypeCollect,
+      InstanceGraph &instanceGraph, DenseSet<Operation *> &overrided,
+      Operation *topOp);
+
+  /// The function to get the number of secondary trigger signal.
+  unsigned getNumOfSecondaryTriggerNum(HWModuleOp op);
+};
+
+struct StateStructGeneratePass
+    : public StateStructGenerateBase<StateStructGeneratePass> {
+  void runOnOperation() override;
+
+  using StateStructGenerateBase<StateStructGeneratePass>::emitNoneInfo;
+};
+} // namespace
+
+bool GenStateStruct::isOutputReg(sv::RegOp op) {
+  bool outputReg = false;
+  for (auto *user : op->getUsers()) {
+    if (isa<hw::OutputOp>(user)) {
+      outputReg = true;
+      break;
+    }
+  }
+  return outputReg;
+}
+
+int GenStateStruct::convertPortDirection(PortInfo &pInfo) {
+  // The value 0 is used to represent the port direction is unknown.
```

Comment:
Isn't this what the enum is for?  LLVM tends to prefer not to have overly defensive programming (internal data structures are always valid).

Reply:
Actually, it's not that necessary, just because I didn't think the enum definition was in PortInfo when I did this part, now it's in ModulePortInfo, I'll change it soon.